app/models/container_label_tag_mapping.rb

```diff
@@ -2,15 +2,99 @@ class ContainerLabelTagMapping < ApplicationRecord
   #   In this case, `tag` specifies the category under which to create
   #   the value-specific tag (and classification) on demand.
   #   We then also add a specific `label_value`->specific `tag` mapping here.
+
   belongs_to :tag
 
-  validates :label_name, :presence => true
+  def self.drop_cache
+    @cached_hash = nil
+    @cached_mapped_tags = nil
+  end
+
+  # Finds all existing specific-value tags we're pointing to.
+  def self.all_mapped_tags
+    @cached_mapped_tags ||= Tag.joins(:container_label_tag_mappings).where.not(
+      :container_label_tag_mappings => {:label_value => nil}
+    ).to_a
+  end
+
+  # Returns {name => {type => {value => [tag, ...]}}} hash.
+  def self.hash_all_by_name_type_value
+    unless @cached_hash
+      @cached_hash = {}
+      includes(:tag).find_each { |m| load_mapping_into_hash(m, @cached_hash) }
+    end
+    @cached_hash
+  end
+
+  def self.load_mapping_into_hash(mapping, hash)
+    name, type, value = mapping.label_name, mapping.labeled_resource_type, mapping.label_value
+    hash[name] ||= {}
+    hash[name][type] ||= {}
+    hash[name][type][value] ||= []
+    hash[name][type][value] << mapping.tag
+  end
+  private_class_method :load_mapping_into_hash
+
+  # Main entry point
+  def self.tags_for_entity(entity)
+    entity.labels.collect_concat { |label| tags_for_label(label) }
+  end
+
+  def self.tags_for_label(label)
+    tags = []
+
+    by_type_value = hash_all_by_name_type_value[label.name] || {}
+    # Apply both specific-type and any-type, independently.
+    [label.resource_type, nil].each do |type|
+      by_value = by_type_value[type] || {}
+      specific = by_value[label.value] || []
+      if !specific.empty?
+        tags.concat(specific)
+      else
+        any_value = by_value[nil] || []
+        any_value.each do |category_tag|
+          # Create a specific-value mapping under same (type-or-nil, name).
+          tags << create_specific_value_mapping(type, label.name, label.value, category_tag).tag
+        end
+      end
+    end
+
+    tags
+  end
+
+  # If this is an open ended any-value mapping, finds or creates a
+  # specific-value mapping to a specific tag.
+  def self.create_specific_value_mapping(type, name, value, category_tag)
+    new_tag = create_tag(category_tag, value)
+    new_mapping = create!(:labeled_resource_type => type, :label_name => name, :label_value => value,
+                          :tag => new_tag)
+    load_mapping_into_hash(new_mapping, @cached_hash) if @cached_hash
+    @cached_mapped_tags << new_tag if @cached_mapped_tags
+    new_mapping
+  end
+  private_class_method :create_specific_value_mapping
+
+  def self.create_tag(category_tag, value)
+    entry_name = Classification.sanitize_name(value)
+    category = category_tag.classification
+    if category
+      entry = category.add_entry(:name => entry_name, :description => value)
+      entry.save!
+      entry.tag
+    else
```

Comment:
I assume a mapping always has a non-null tag_id.  It's completely pointless otherwise.  (Though there is no null: false DB constaint on it.)
Tags may exist without a classfication.  I don't know how much that actually happens.  I could (1) merely document that I only deal with tags that have a classification (2) also detect and signal failure (3) just deal with any tag.  Even formulating (1) that proved messier than (3)...

[I'm not creating an entry Classification in this branch because entries can't have parent_id = 0.]